README.md
Pix-Diff
==========

A lightweight protractor plugin for image comparison

[![Build Status](https://travis-ci.org/koola/pix-diff.svg)](https://travis-ci.org/koola/pix-diff)
[![npm version](https://badge.fury.io/js/pix-diff.svg)](http://badge.fury.io/js/pix-diff)

[![NPM](https://nodei.co/npm/pix-diff.png)](https://nodei.co/npm/pix-diff/)

##Installation

Install this module locally with the following command:
```shell
npm install pix-diff
```

Save to dependencies or dev-dependencies:
```shell
npm install --save pix-diff
npm install --save-dev pix-diff
```

##Usage

The package can be used directly in individual tests or via ```onPrepare``` in the Protractor configuration file.

**Example:**
```javascript
exports.config = {
   // your config here ...

    onPrepare: function() {
        var PixDiff = require('pix-diff');
        browser.pixdiff = new PixDiff(
            {
                basePath: 'path/to/screenshots/',
                width: 1280,
                height: 1024
            }
        );
    },
}
```

PixDiff provides two comparison methods ```checkScreen``` and ```checkRegion``` along with two jasmine matchers ```toMatchScreen``` and ```toNotMatchScreen```. Two helper methods ```saveScreen``` and ```saveRegion``` are provided for saving images.

**Example:**
```javascript
describe("Example page", function() {

    beforeEach(function() {
        browser.get('http://www.example.com/');
    });

    it("should match the page", function () {
        expect(browser.pixdiff.checkScreen('example-page')).toMatchScreen();
    });

    it("should not match the page", function () {
        element(By.buttonText('yes')).click();
        expect(browser.pixdiff.checkScreen('example-page')).toNotMatchScreen();
    });

    it("should match the title", function () {
        expect(browser.pixdiff.checkRegion(element(By.id('title')), 'example-page-title')).toMatchScreen();
    });

    it("should match the title", function () {
        expect(browser.pixdiff.checkRegion(element(By.id('title')), 'example-page-title', {
            blockOut: {x: 10, y: 132, width: 100, height: 50}})).toMatchScreen();
    });
});
```

####PixDiff Parameters:

* ```basePath``` Defines the path to the reference images that are to be compared.
* ```width``` Browser width (default: 1280)
* ```height``` Browser height (default: 1024)

####Function options:

* ```blockOut``` Object or list of objects with coordinates that should be blocked before comparing. (default: none)
* ```debug``` When set, then block-out regions will be shown on the output image. (default: false)

####Cropping
Images can be cropped before they are compared by using the ```checkRegion``` function. The function will calculate the correct dimensions based upon the webdriver element selector (see example above).

####Block-Out
Sometimes, it is necessary to block-out some specific areas in an image that should be ignored for comparisons. For example, this can be IDs or even time-labels that change with the time. Adding block-outs to images may decrease false positives and therefore stabilizes these comparisons (see example above).

## Conventions
There are directory and naming conventions that must be met.

**Directory structure**
```text
path
└── to
    └── screenshots
        ├── diff
        │   └── example-page-chrome-1280x1024.png
        ├── example-page-chrome-800x600.png
        ├── example-page-chrome-1280x1024.png
        ├── example-page-title-chrome-800x600.png
        └── example-page-title-chrome-1280x1024.png
```
The ```basePath``` directory must contain all the *approved* images. You may create subdirectories for better organisation, but the relative path should then be given in the test spec method. Failed comparisons generate a diff image under the **diff** folder.

**Image naming**

Images should obey the following format:

```text
{descriptive_name}-{browserName}-{browserWidth}x{browserHeight}.png
```
Images specified via name in the spec method will be selected according to the browsers current resolution. That is to say that multiple images can share the same name differentated by resolution.

##Documentation

Todo

##Tests

Run the tests with the following command:
```shell
npm run test
```

###Dependencies
* [blink-diff](https://github.com/yahoo/blink-diff)
* [png-image](https://github.com/koola/png-image)

###Dev-Dependencies
* [grunt](https://github.com/gruntjs/grunt)
* [grunt-cli](https://github.com/gruntjs/grunt-cli)
* [grunt-contrib-clean](https://github.com/gruntjs/grunt-contrib-clean)
* [grunt-protractor-runner](https://github.com/teerapap/grunt-protractor-runner)
* [load-grunt-tasks](https://github.com/sindresorhus/load-grunt-tasks)
* [protractor](https://github.com/angular/protractor)

##License

The MIT License

Copyright 2016 Koola.

CHANGELOG.md
Changelog
=========
v1.0.7 - Merged PR #5 to resolve issue on SauceLabs (18/12/2015)

v1.0.6 - Rename custom matchers to avoid naming conflict with Jasmine toMatch (04/11/2015)

v1.0.5 - Add Jasmine 2 custom matchers (10/07/2015)

v1.0.4 - Fix writing diff when sub-directories are under base path (30/12/2014)

v1.0.3 - Added tests (28/12/2014)

v1.0.2 - Added image save methods (17/12/2014)

v1.0.1 - Add missing percentage sign and documentation (12/12/2014)

v1.0.0 - Initial release (11/12/2014)

test/protractor.conf.js
'use strict';

exports.config = {

    baseUrl: 'http://www.example.com',

    specs: ['*.spec.js'],

    capabilities: {
        browserName: 'chrome',
        chromeOptions: {
            args: ['--no-sandbox']
        }
    },

    onPrepare: function() {
        var PixDiff = require('../');
        browser.pixDiff = new PixDiff({
            basePath: 'test/screenshots',
            width: 800,
            height: 600
        });

        browser.ignoreSynchronization = true;
    },

    jasmineNodeOpts: {
        defaultTimeoutInterval: 60000,
        showColors: true
    }
};

test/integration.spec.js
'use strict';

var BlinkDiff = require('blink-diff'),
    fs = require('fs');

describe("Pix-Diff", function() {

    beforeEach(function() {
        browser.get(browser.baseUrl);
    });

    it("should save the screen", function () {
        var tagName = 'example-page';

        browser.pixDiff.saveScreen(tagName).then(function() {
            expect(fs.existsSync(__dirname + '/screenshots/' + tagName + '-chrome-800x600.png')).toBe(true);
        });
    });

    it("should match the page", function () {
        browser.pixDiff.checkScreen('example-page').then(function(result) {
            expect(result.code).toEqual(BlinkDiff.RESULT_IDENTICAL);
        });
    });

    it("should match the page with custom matcher", function () {
        expect(browser.pixDiff.checkScreen('example-page')).toMatch();
    });

    it("should not match the page", function () {
        browser.pixDiff.checkScreen('example-fail', {threshold:1}).then(function(result) {
            expect(result.code).toEqual(BlinkDiff.RESULT_DIFFERENT);
        });
    });

    it("should not match the page with custom matcher", function () {
        expect(browser.pixDiff.checkScreen('example-fail', {threshold:1})).toNotMatch();
    });
});

index.js
var BlinkDiff = require('blink-diff'),
    PNGImage = require('png-image'),
    assert = require('assert'),
    path = require('path'),
    util = require('util');

/**
 * Pix-diff protractor plugin class
 *
 * @constructor
 * @class PixDiff
 * @param {object} options
 * @param {string} options.basePath Path to screenshots folder
 * @param {string} options.width Width of browser
 * @param {string} options.height Height of browser
 *
 * @property {string} _basePath
 * @property {int} _width
 * @property {int} _height
 * @property {object} _capabilities
 * @property {webdriver|promise} _flow
 */
function PixDiff(options) {
    this._basePath = options.basePath;
    assert.ok(options.basePath, "Image base path not given.");

    this._width = options.width || 1280;
    this._height = options.height || 1024;

    this._capabilities = null;

    this._flow = browser.controlFlow();

    // init
    browser.driver.manage().window().setSize(this._width, this._height)
        .then(function() {
            return browser.getCapabilities()
        })
        .then(function(data) {
            return this._capabilities = data.caps_;
        }.bind(this));
}

PixDiff.prototype = {

    /**
     * Merges non-default options from optionsB into optionsA
     *
     * @method _mergeDefaultOptions
     * @param {object} optionsA
     * @param {object} optionsB
     * @return {object}
     * @private
     */
    _mergeDefaultOptions: function (optionsA, optionsB) {
        var option;

        optionsB = (typeof optionsB === 'object') ? optionsB : {};

        for (option in optionsB) {
            if (!optionsA.hasOwnProperty(option)) {
                optionsA[option] = optionsB[option];
            }
        }
        return optionsA;
    },

    /**
     * Saves an image of the screen
     *
     * @method saveScreen
     * @example
     *     browser.pixdiff.saveScreen('imageA');
     *
     * @param {string} tag
     * @public
     */
    saveScreen: function(tag) {
        return this._flow.execute(function() {
            return browser.takeScreenshot()
                .then(function(image) {
                    tag = util.format('%s-%s-%sx%s.png', tag, this._capabilities.browserName, this._width, this._height);
                    return new PNGImage({
                        imagePath: new Buffer(image, 'base64'),
                        imageOutputPath: path.join(this._basePath, tag)
                    }).runWithPromise();
                }.bind(this));
        }.bind(this));
    },

    /**
     * Saves an image of the screen region
     *
     * @method saveRegion
     * @example
     *     browser.pixdiff.saveRegion(element(By.id('elementId')), 'imageA');
     *
     * @param {promise} element
     * @param {string} tag
     * @public
     */
    saveRegion: function(element, tag) {
        var size,
            rect;

        return this._flow.execute(function() {
            return element.getSize()
                .then(function(elementSize) {
                    size = elementSize;
                    return element.getLocation();
                })
                .then(function(point) {
                    rect = {height: size.height, width: size.width, x: Math.floor(point.x), y: Math.floor(point.y)};
                    return browser.takeScreenshot();
                })
                .then(function(image) {
                    tag = util.format('%s-%s-%sx%s.png', tag, this._capabilities.browserName, this._width, this._height);
                    return new PNGImage({
                        imagePath: new Buffer(image, 'base64'),
                        imageOutputPath: path.join(this._basePath, tag),
                        cropImage: rect
                    }).runWithPromise();
                }.bind(this));
        }.bind(this));
    },

    /**
     * Runs the comparison against the screen
     *
     * @method checkScreen
     * @example
     *     browser.pixdiff.checkScreen('imageA', {debug: true});
     *
     * @param {string} tag
     * @param {object} options
     * @return {object} result
     * @public
     */
    checkScreen: function(tag, options) {
        var defaults;

        return this._flow.execute(function() {
            return browser.takeScreenshot()
                .then(function(image) {
                    tag = util.format('%s-%s-%sx%s.png', tag, this._capabilities.browserName, this._width, this._height);
                    defaults = {
                        imageAPath: path.join(this._basePath, tag),
                        imageB: new Buffer(image, 'base64'),
                        imageOutputPath: path.join(this._basePath, 'diff', path.basename(tag)),
                        imageOutputLimit: BlinkDiff.OUTPUT_DIFFERENT
                    };
                    return new BlinkDiff(this._mergeDefaultOptions(defaults, options)).runWithPromise();
                }.bind(this))
                .then(function(result) {
                    return result;
                });
        }.bind(this));
    },

    /**
     * Runs the comparison against a region
     *
     * @method checkRegion
     * @example
     *     browser.pixdiff.checkRegion(element(By.id('elementId')), 'imageA', {debug: true});
     *
     * @param {promise} element
     * @param {string} tag
     * @param {object} options
     * @return {object}
     * @public
     */
    checkRegion: function(element, tag, options) {
        var size,
            rect,
            defaults;

        return this._flow.execute(function() {
            return element.getSize()
                .then(function(elementSize) {
                    size = elementSize;
                    return element.getLocation();
                })
                .then(function(point) {
                    rect = {height: size.height, width: size.width, x: Math.floor(point.x), y: Math.floor(point.y)};
                    return browser.takeScreenshot();
                })
                .then(function(image) {
                    tag = util.format('%s-%s-%sx%s.png', tag, this._capabilities.browserName, this._width, this._height);
                    defaults = {
                        imageAPath: path.join(this._basePath, tag),
                        imageB: new Buffer(image, 'base64'),
                        imageOutputPath: path.join(this._basePath, 'diff', path.basename(tag)),
                        imageOutputLimit: BlinkDiff.OUTPUT_DIFFERENT,
                        cropImageB: rect
                    };
                    return new BlinkDiff(this._mergeDefaultOptions(defaults, options)).runWithPromise();
                }.bind(this))
                .then(function(result) {
                    return result;
                });
        }.bind(this));
    }
};

/**
 * Jasmine PixDiff matchers
 */
(function() {
    var v1 = {
            toMatchScreen: function() {
                var result = this.actual,
                    percent = +((result.differences / result.dimension) * 100).toFixed(2);
                this.message = function() {
                    return util.format("Image is visibly different by %s pixels, %s %", result.differences, percent);
                };
                return ((result.code === BlinkDiff.RESULT_IDENTICAL) || (result.code === BlinkDiff.RESULT_SIMILAR));
            },

            toNotMatchScreen: function() {
                var result = this.actual;
                this.message = function() {
                    return "Image is identical or near identical";
                };
                return ((result.code === BlinkDiff.RESULT_DIFFERENT) && (result.code !== BlinkDiff.RESULT_UNKNOWN));
            }
        },
        v2 = {
            toMatchScreen: function() {
                return {
                    compare: function(actual, expected) {
                        var percent = +((actual.differences / actual.dimension) * 100).toFixed(2);
                        return {
                            pass: ((actual.code === BlinkDiff.RESULT_IDENTICAL) || (actual.code === BlinkDiff.RESULT_SIMILAR)),
                            message: util.format("Image is visibly different by %s pixels, %s %", actual.differences, percent)
                        };
                    }
                }
            },

            toNotMatchScreen: function() {
                return {
                    compare: function(actual, expected) {
                        return {
                            pass: ((actual.code === BlinkDiff.RESULT_DIFFERENT) && (actual.code !== BlinkDiff.RESULT_UNKNOWN)),
                            message: "Image is identical or near identical"
                        };
                    }
                }
            }
        };

    beforeEach(function() {
        if (/^2/.test(jasmine.version)) {
            jasmine.addMatchers(v2);
        } else {
            this.addMatchers(v1);
        }
    });
})();

module.exports = PixDiff;